src/impl_/pymodule.rs

```diff
@@ -12,7 +14,10 @@ pub struct ModuleDef {
     // wrapped in UnsafeCell so that Rust compiler treats this as interior mutability
     ffi_def: UnsafeCell<ffi::PyModuleDef>,
     initializer: ModuleInitializer,
-    initialized: AtomicBool,
+    #[cfg(once_lock)]
+    interpreter: OnceLock<i64>,
```

Comment:
From https://docs.python.org/3/c-api/init.html#c.PyInterpreterState_GetID, I think we can use `-1` as a manual niche, i.e. just turn this into a `AtomicI64` initialized to `-1` and considered uninitialized as long as it has this value. This should be faster and simpler than both `OnceLock<i64>` and `Mutex<Option<i64>>`.

Reply:
In https://github.com/pydantic/pydantic/issues/6584#issuecomment-1673260886, you mention opt-in. Might this be advisable even for this limited relaxation? What if the module is torn down but `static` data still contains pointers into the old module (e.g. capsules) which are now invalid? For example, is the code in https://github.com/PyO3/rust-numpy/blob/c16fbb1e630b0538fdf17cbc53043bf0467459f9/src/npyffi/mod.rs#L30-L32 valid after this change?

Reply:
There is `PyErr_Occurred` (or other APIs which read the error indicator) which can be used to disambiguate -1 return values between success and error cases. For example `PyLong_AsLong`.

I think in this case you're right that interpreter IDs are implemented by a counter starting at 0 and then increment upwards. There is always the worry that this changes in the future, but hey, maybe by then we'll have an MSRV with `OnceLock` stable 😝 

I note also that I haven't actually handled the -1 error case here, so I need to correct for that too.